src/resolve/resolvectl.c

```diff
@@ -66,29 +69,179 @@ typedef enum StatusMode {
+
+        while ((r = sd_bus_message_read(reply, "(iso)", &ifindex, &ifname, &path)) > 0) {
+                _cleanup_(link_freep) Link *l = NULL;
+                _cleanup_free_ char *n = NULL, *p = NULL;
+
+                if (ifindex <= 0 || isempty(ifname) || isempty(path))
+                        continue;
+
+                n = strdup(ifname);
+                if (!n)
+                        return log_oom();
+
+                p = strdup(path);
+                if (!n)
+                        return log_oom();
+
+                l = new(Link, 1);
+                if (!l)
+                        return log_oom();
+
+                *l = (Link) {
+                        .ifindex = ifindex,
+                        .ifname = TAKE_PTR(n),
+                        .path = TAKE_PTR(p),
+                };
+
+                LIST_APPEND(links, links, TAKE_PTR(l));
+        }
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        return 0;
+}
+
+static int link_index_to_name(int ifindex, char **ret) {
+        char *ifname = NULL;
+
+        assert(ret);
+
+        if (ifindex <= 0)
+                return -EINVAL;
+
+        if (arg_transport == BUS_TRANSPORT_LOCAL) {
+                char name[IF_NAMESIZE] = {};
+
+                if (!if_indextoname((unsigned) ifindex, name))
+                        return -errno;
+
+                ifname = strdup(name);
+                if (!ifname)
+                        return -ENOMEM;
+
+        } else {
+                Link *l;
+
+                LIST_FOREACH(links, l, links)
+                        if (ifindex == l->ifindex) {
+                                ifname = strdup(l->ifname);
+                                if (!ifname)
+                                        return -ENOMEM;
+
+                                break;
+                        }
+
+                if (!ifname)
+                        return -ENODEV;
+        }
+
+        *ret = TAKE_PTR(ifname);
+        return 0;
+}
+
 static int parse_ifindex_and_warn(const char *s) {
-        int ifi;
+        int ifindex;
+        Link *l;
 
         assert(s);
 
-        if (parse_ifindex(s, &ifi) < 0) {
-                ifi = if_nametoindex(s);
-                if (ifi <= 0)
-                        return log_error_errno(errno, "Unknown interface '%s': %m", s);
-        }
+        if (parse_ifindex(s, &ifindex) < 0) {
+                /* input seems to be a ifname */
+                if (arg_transport == BUS_TRANSPORT_LOCAL) {
+                        ifindex = if_nametoindex(s);
+                        if (ifindex <= 0)
+                                return log_error_errno(errno, "Unknown interface '%s': %m", s);
+
+                        return ifindex;
+                } else {
+                        LIST_FOREACH(links, l, links)
+                                if (streq_ptr(l->ifname, s))
+                                        return l->ifindex;
```

Comment:
instead of iterating through the list here, why not add a call a pair of calls GetLinkByName() and GetLinkByIndex() that return object path and the matching other value?